pkg/scheduler/plugins/gpushare/plugin.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	nodeFreeGPUResource := nodeDeviceInfo.deviceFree[schedulingv1alpha1.GPU]
+
+	if checkNodeGPUSatisfied(podRequest, nodeFreeGPUResource) {
+		return nil
+	}
+
+	return framework.NewStatus(framework.Unschedulable, ErrNoMatchNode)
+}
+
+func (g *Plugin) Reserve(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	state, err := getPreFilterState(cycleState)
+	if err != nil {
+		return framework.AsStatus(err)
+	}
+	if state.skip {
+		return nil
+	}
+
+	nodeInfo, err := g.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	nodeDeviceInfo := g.nodeDeviceCache.getNodeDevice(node.Name)
+	if nodeDeviceInfo == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrMissingDevice)
+	}
+
+	nodeDeviceInfo.lock.Lock()
+	defer nodeDeviceInfo.lock.Unlock()
+	nodeDeviceTotal := nodeDeviceInfo.deviceTotal[schedulingv1alpha1.GPU]
+	// node does not have GPU
+	if len(*nodeDeviceTotal) <= 0 {
+		return framework.NewStatus(framework.Unschedulable, ErrNoMatchNode)
+	}
+
+	podRequest := state.convertedResource
+
+	if gpuMem, ok := podRequest[apiext.GPUMemory]; ok {
+		podRequest[apiext.GPUMemoryRatio] = memBytesToRatio(gpuMem, (*nodeDeviceTotal)[0][apiext.GPUMemory])
+	} else {
+		gpuMemRatio := podRequest[apiext.GPUMemoryRatio]
+		podRequest[apiext.GPUMemory] = memRatioToBytes(gpuMemRatio, (*nodeDeviceTotal)[0][apiext.GPUMemory])
+	}
+
+	nodeDeviceFree := nodeDeviceInfo.deviceFree[schedulingv1alpha1.GPU]
+
+	allocateResult := allocateGPU(podRequest, nodeDeviceFree)
+
+	g.nodeDeviceCache.assume(allocateResult, nil, nodeName)
+
+	state.allocResult = allocateResult
+	return nil
+}
+
+func (g *Plugin) Unreserve(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	state, err := getPreFilterState(cycleState)
+	if err != nil {
+		return
+	}
+	if state.skip {
+		return
+	}
+	nodeDeviceInfo := g.nodeDeviceCache.getNodeDevice(nodeName)
+	if nodeDeviceInfo == nil {
+		return
+	}
+
+	g.nodeDeviceCache.unAssume(state.allocResult, nil, nodeName)
+
+	state.allocResult = nil
+}
+
+func (g *Plugin) PreBind(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	state, err := getPreFilterState(cycleState)
+	if err != nil {
+		return framework.AsStatus(err)
+	}
+	if state.skip {
+		return nil
+	}
+
+	allocResult := state.allocResult
+	newPod := pod.DeepCopy()
+	apiext.SetGPUsAllocation(newPod, allocResult)
+	patchBytes, err := util.GeneratePodPatch(pod, newPod)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	err = retryOnConflictOrTooManyRequests(func() error {
```

Comment:
if a GPU pod fails to schedule after this Patch, should the plugin undoes the Patch

Reply:
Nice. I'll leave a TODO here, and fix it in the next PR.